build-system/compile/json-schema/index.js

```diff
@@ -182,42 +182,209 @@ function transformAjvCode(code, scope, config) {
+      } else if (path.isIdentifier()) {
+        const binding = path.scope.getBinding(path.node.name);
+        if (!binding) {
+          break;
+        }
+        path = binding.path;
+      } else {
+        break;
+      }
+    }
+    return i === 0 && path.isIdentifier({name: keys[i]});
+  }
+
+  /**
+   * @param {babel.NodePath} path
+   * @return {boolean}
+   */
+  function isReferenceToObjectKeyList(path) {
+    const {parentPath} = path;
+    if (parentPath?.isMemberExpression() && path.key === 'object') {
+      return (
+        t.isIdentifier(parentPath.node.property, {name: 'hasOwnProperty'}) &&
+        parentPath.parentPath?.isCallExpression()
+      );
+    }
+    if (parentPath?.isCallExpression() && path.key === 0) {
+      const callee = parentPath.get('callee');
+      return (
+        isMemberExpressionLeftwards(callee, ['Object', 'keys']) ||
+        isMemberExpressionLeftwards(callee, [
+          'Object',
+          'prototype',
+          'hasOwnProperty',
+          'call',
+        ])
+      );
+    }
+    return false;
+  }
+
+  /** @type {{[stringified: string]: string}} */
+  const valueReferences = {};
+
+  /**
+   * Resolves static references to object properties as deeply as possible,
+   * and hoists references to those values.
+   * @param {import('@babel/traverse').Binding} binding
+   */
+  function hoistProperties(binding) {
+    const {constant, identifier, path, referencePaths} = binding;
+    if (!constant || !path.isVariableDeclarator()) {
+      return;
+    }
+    const init = path.get('init');
+    if (!init.isObjectExpression()) {
+      return;
+    }
+    const original = init.evaluate().value;
+    if (original == null) {
+      // original is dynamic, deopt
+      return;
+    }
+    /** @type {Map<Object, babel.NodePath[]>} */
+    const hoisted = new Map();
+    /** @type {[any, babel.NodePath[]][]} */
+    const queue = [[original, referencePaths]];
+    for (const [startValue, referencePaths] of queue) {
+      for (let referencePath of referencePaths) {
+        let value = startValue;
+        let {parentPath} = referencePath;
+        while (parentPath?.isMemberExpression()) {
+          const key = evaluatePropertyKey(parentPath);
+          if (key == null || !value.hasOwnProperty(key)) {
+            break;
+          }
+          value = value[key];
+          referencePath = parentPath;
+          parentPath = referencePath.parentPath;
+        }
+        if (value === original) {
+          // deopt
+          return;
+        }
+        // Follow references to declared variable.
+        if (
+          parentPath?.isVariableDeclarator() &&
+          t.isIdentifier(parentPath.node.id)
+        ) {
+          const {name} = parentPath.node.id;
+          const binding = parentPath.scope.getBinding(name);
+          if (binding?.constant) {
+            const {referencePaths} = binding;
+            queue.push([value, referencePaths]);
+            parentPath.remove();
```

Comment:
Can this leave us in a bad state? Eg, during the next loop we hit a deopt.

If we left this code in place, would the DCE remove it later when doing static analysis.